### ISSUE TITLE
authentication tokens from npmrc not working

### ISSUE THREAD

**Issue by lookfirst** [jspm/npm#78]
```
jon:hub jon$ jspm registry config npm
Currently using a custom registry. Would you like to reset to npmrc defaults? [no]:yes
npm registry [https://registry.npmjs.org]:
Currently reading credentials from npmrc, configure custom authentication? [no]:yes
Currently using a username and password. Configure token-based authentication? [no]:
Enter your npm username [lookfirst]:
Enter your npm password:
Would you like to test these credentials? [yes]:
ok   npm authentication is working successfully.
ok   Registry npm configured successfully.
jon:hub jon$ jspm install
     Looking up npm:babel-core
     Looking up npm:babel-runtime
     Looking up npm:core-js
     Looking up npm:lodash
     Updating registry cache...

warn Error on lookup for npm:babel-runtime
     Invalid authentication details. Run jspm registry config npm to reconfigure.

err  Error looking up npm:babel-runtime.

warn Installation changes not saved.
jon:hub jon$ jspm --version
0.16.0-beta
Running against local jspm install.
jon:hub jon$
```

**Comment by guybedford** [comment]
Sure I'll look into this.

**Comment by lookfirst** [comment]
I ended up deleting the npm section from the config file and starting over without auth. That seemed to fix the issue. No idea what it was since I hadn't touched that section before. It only started after 'upgrading' to .16 beta.

**Comment by jimschubert** [comment]
@guybedford 

Looks like the problem is with the default registry being set to `https://registry.npmjs.org` [here](https://github.com/jspm/npm/blob/1fb82279b46c833e4e0d7c5b5a879bb478dcc7e7/lib/npm.js#L15). This is used in getting the auth on line 56 of that same file.

I'm using npm 2.10.1 currently, but I'm not sure what version I was using when I created my `~/.npmrc`. The npm module in jspm assumes the default registry is `https://registry.npmjs.org`, but npm enters the registry as `//registry.npmjs.org`. I *think* this causes `getOption` in npmrc.js to fail on the regex for the prefix (or later in getOptionExact):

```
function getOption(contents, key, registry) {
  if (registry) {
    var prefix = registry.replace(/^.+:/, '') + '/:';
    return getOptionExact(contents, prefix + key) || getOptionExact(contents, key);
  }
  else
    return getOptionExact(contents, key);
}
```

Setting the registry settings in `~/.npmrc` with the prefix of `https:` allows jspm to install npm packages. I don't think that will work correctly with npm, though. For now, my workaround is backing up .npmrc with its original settings while using 0.16 beta.

I did a quick search, and [npm's tests](https://github.com/npm/npm/blob/180da67c9fa53103d625e2f031626c2453c7ebcd/test/tap/map-to-registry.js#L7) have an example of mapping the protocol relative URL to the registry and extracting auth, username, etc. The format of those settings look pretty complex, especially with registry scoping.

**Comment by guybedford** [comment]
@lookfirst @jimschubert without sharing your tokens, could you perhaps share the structure of your npmrc files here? It would really help for testing.

@lookfirst I'm not sure how you are getting that issue at all, are you sure the details are correct? If you log `this.auth` just before https://github.com/jspm/npm/blob/master/lib/npm.js#L57, can you check if the token has been read correctly?

@jimschubert in my test case I exactly have `https://registry.npmjs.org` as the registry value and it's looking up the token from `//registry.npmjs.org/:_authToken=...` in the npmrc fine.

**Comment by guybedford** [comment]
Unfortunately I've been unable to replicate any of these issues without seeing the npmrc files in question.

**Comment by jimschubert** [comment]
This is my unmodified `~/.npmrc`:

```
//registry.npmjs.org/:_password=passwordhere
//registry.npmjs.org/:username=jimschubert
//registry.npmjs.org/:email=james.schubert@gmail.com
//registry.npmjs.org/:always-auth=false
```

I don't have an `_authToken` setting, but I don't need one with jspm 0.15.7. Maybe I'm missing some step in the upgrade to 0.16 beta.

**Comment by guybedford** [comment]
@jimschubert that case is working ok for me actually. Is the password entry encoded at all?

**Comment by guybedford** [comment]
Note make sure you run `jspm registry config npm` and select reset to npmrc defaults.

**Comment by lookfirst** [comment]
I torched the failed file. I'm sorry. =(

**Comment by lookfirst** [comment]
That said, I keep going back to this whole thing of replicating the `npm` cli functionality in jspm. Let's drop that. ;-)

**Comment by jimschubert** [comment]
Yeah my password is encoded.

Here's what I get when running `jspm registry config npm`:

```
jim at macbert in ~/temp/electron-aurelia-example on gulp-webdriver* e19fcef
$ jspm registry config npm
Currently using a custom registry. Would you like to reset to npmrc defaults? [no]:yes
npm registry [https://registry.npmjs.org]:
Currently reading credentials from npmrc, configure custom authentication? [no]:yes
Configure token-based authentication? [no]:
Enter your npm username:jimschubert
Enter your npm password:
Would you like to test these credentials? [yes]:

err  Error: options.uri is a required argument
         at Request.init (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/jspm-npm/node_modules/request/request.js:358:31)
         at new Request (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/jspm-npm/node_modules/request/request.js:264:8)
         at request (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/jspm-npm/node_modules/request/index.js:50:10)
         at lib$rsvp$node$$tryApply (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1467:11)
         at lib$rsvp$node$$handleValueInput (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1567:20)
         at fn (/Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/rsvp/dist/rsvp.js:1555:18)
         at /Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/jspm-npm/lib/auth.js:76:26
         at runMicrotasksCallback (node.js:337:7)
         at process._tickCallback (node.js:355:11)

Would you like to try new credentials? [yes]:no

err  TypeError: Cannot read property 'token' of undefined
         at /Users/jim/temp/electron-aurelia-example/node_modules/jspm/node_modules/jspm-npm/lib/npm.js:106:16
         at runMicrotasksCallback (node.js:337:7)
         at process._tickCallback (node.js:355:11)
jim at macbert in ~/temp/electron-aurelia-example on gulp-webdriver* e19fcef
$ jspm --version
0.16.0-beta
Running against local jspm install.
```

I figured out that these issues only seem to happen when I'm not logged into npm.  Running `npm login` generates the single line in `~/.npmrc`:

```
//registry.npmjs.org/:_authToken=TOKEN
```

where TOKEN is my actual token. Then, running `jspm install` on 0.16 beta completes successfully.

**Comment by jimschubert** [comment]
More info that may be relevant:

```
$ npm --version
2.10.1
$ node --version
v0.12.3
```

I'm running OS X Yosemite 10.10.3. I run node through nave and 0.12.3 is set as my main.

**Comment by jimschubert** [comment]
I found a hint from #74 and the pull request I've just opened fixes the issue for me when I'm not logged in, and using the `~/.npmrc` I posted previously.

I don't know why NPM's registry would only work with bearer authentication now, though. Maybe it's an issue with NPM load balancing and one of the servers is misconfigured?

```
$ ping -t 1 registry.npmjs.org
PING a.sni.fastly.net (23.235.39.162): 56 data bytes
64 bytes from 23.235.39.162: icmp_seq=0 ttl=58 time=24.548 ms

--- a.sni.fastly.net ping statistics ---
1 packets transmitted, 1 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 24.548/24.548/24.548/0.000 ms
```

**Comment by guybedford** [comment]
@jimschubert thanks - those two bugs you just posted are fixed in https://github.com/jspm/npm/commit/d9626cae6232a21df5a2c1c49eab70602b596d24. Let me know if you have an update on the other point.

**Comment by guybedford** [comment]
Included a fix in https://github.com/jspm/npm/commit/77a11e85ffe828104587526a2d1068f30c83e9d7 which is a big one. There may still be other edge cases but closing for now as I think we've covered the majority now.

**Comment by jimschubert** [comment]
After your fix for the two bugs in d9626ca, I wasn't able to repro the behavior I was seeing before.

**Comment by guybedford** [comment]
Added some more fixes here in https://github.com/jspm/npm/commit/94b92299fd8521d8121cddeb9ac91edf1669343f and released in 0.23.3, so this feature should finally be getting stable. Thanks to @harrison for helping to track this down.
